sql/analyzer/apply_foreign_keys.go

```diff
@@ -160,57 +165,249 @@ func applyForeignKeysToNodes(ctx *sql.Context, a *Analyzer, n sql.Node, cache *f
 	}
 }
 
+// getForeignKeyEditor merges both getForeignKeyReferences and getForeignKeyRefActions and returns a single editor.
+func getForeignKeyEditor(ctx *sql.Context, a *Analyzer, tbl sql.ForeignKeyTable, cache *foreignKeyTableCache, fkChain foreignKeyChain) (*plan.ForeignKeyEditor, error) {
+	fkEditor, err := getForeignKeyReferences(ctx, a, tbl, cache, fkChain)
+	if err != nil {
+		return nil, err
+	}
+	fkEditor2, err := getForeignKeyRefActions(ctx, a, tbl, cache, fkChain)
+	if err != nil {
+		return nil, err
+	}
+	// Due to tables not containing their database, we have to grab the table's database from the foreign key definition.
+	// This is so we may make proper use of cached updaters. Therefore, both calls return a full editor rather than just
+	// their respective specialties (and may be used independently).
+	if fkEditor == nil && fkEditor2 == nil {
+		return nil, nil
+	} else if fkEditor != nil && fkEditor2 != nil {
+		fkEditor.RefActions = fkEditor2.RefActions
+		return fkEditor, nil
+	} else if fkEditor != nil {
+		return fkEditor, nil
+	} else {
+		return fkEditor2, nil
+	}
+}
+
+// getForeignKeyReferences returns an editor containing only the references for the given table.
+func getForeignKeyReferences(ctx *sql.Context, a *Analyzer, tbl sql.ForeignKeyTable, cache *foreignKeyTableCache, fkChain foreignKeyChain) (*plan.ForeignKeyEditor, error) {
+	var editor sql.ForeignKeyUpdater
+	declaredFks, err := tbl.GetDeclaredForeignKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// We can ignore foreign keys that have been previously used as we know that the data we're referencing is being inserted
```

Comment:
If we've already added a foreign key to the `fkChain`, then we don't need to validate that its parent table contains the rows we're inserting?

Reply:
Correct, as this child table would only be called as a result of a cascading referential action, so we can guarantee it's in the parent table.